// @wdalmut/opentelemetry-plugin-mongoose/lib/src/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AttributeNames;
(function (AttributeNames) {
    AttributeNames["COMPONENT"] = "component";
    AttributeNames["DB_TYPE"] = "db.type";
    AttributeNames["DB_NAME"] = "db.name";
    AttributeNames["DB_HOST"] = "db.host";
    AttributeNames["DB_PORT"] = "db.port";
    AttributeNames["DB_USER"] = "db.user";
    AttributeNames["COLLECTION_NAME"] = "db.collection";
    AttributeNames["DB_STATEMENT"] = "db.statement";
    AttributeNames["DB_OPTIONS"] = "db.options";
    AttributeNames["DB_UPDATE"] = "db.updates";
    AttributeNames["DB_SAVE"] = "db.save";
    AttributeNames["DB_RESPONSE"] = "db.response";
    AttributeNames["MONGO_ERROR_CODE"] = "db.error_code";
    AttributeNames["DB_MODEL"] = "db.doc";
    AttributeNames["DB_MODEL_NAME"] = "mongoose.model";
    AttributeNames["DB_QUERY_TYPE"] = "mongoose.query";
    AttributeNames["DB_AGGREGATE_PIPELINE"] = "mongoose.db.aggregate_pipeline";
})(AttributeNames = exports.AttributeNames || (exports.AttributeNames = {}));
